src/secp256k1.c

```diff
@@ -785,6 +792,74 @@ int secp256k1_xonly_pubkey_serialize(const secp256k1_context* ctx, unsigned char
     return 1;
 }
 
+int secp256k1_xonly_pubkey_from_pubkey(const secp256k1_context* ctx, secp256k1_xonly_pubkey *xonly_pubkey, int *sign, const secp256k1_pubkey *pubkey) {
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(xonly_pubkey != NULL);
+    ARG_CHECK(pubkey != NULL);
+
+    *xonly_pubkey = *(secp256k1_xonly_pubkey *) pubkey;
+
+    secp256k1_ec_pubkey_absolute(ctx, (secp256k1_pubkey *) xonly_pubkey, sign);
+    return 1;
+}
+
+int secp256k1_xonly_pubkey_to_pubkey(const secp256k1_context* ctx, secp256k1_pubkey *pubkey, const secp256k1_xonly_pubkey *xonly_pubkey, int sign) {
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(xonly_pubkey != NULL);
+
+    *pubkey = *(secp256k1_pubkey *) xonly_pubkey;
+    if (sign) {
+        return secp256k1_ec_pubkey_negate(ctx, pubkey);
+    }
+    return 1;
+}
+
+int secp256k1_xonly_privkey_tweak_add(const secp256k1_context* ctx, unsigned char *seckey32, const unsigned char *tweak32) {
+    secp256k1_ge ge;
+    secp256k1_pubkey pubkey;
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(seckey32 != NULL);
+    ARG_CHECK(tweak32 != NULL);
+
+    if (!secp256k1_ec_pubkey_create(ctx, &pubkey, seckey32)) {
+        return 0;
+    }
+    secp256k1_pubkey_load(ctx, &ge, &pubkey);
+    if (!secp256k1_fe_is_quad_var(&ge.y)) {
+        if (!secp256k1_ec_privkey_negate(ctx, seckey32)) {
+            return 0;
+        }
+    }
+
+    return secp256k1_ec_privkey_tweak_add(ctx, seckey32, tweak32);
+}
+
+int secp256k1_xonly_pubkey_tweak_add(const secp256k1_context* ctx, secp256k1_pubkey *output_pubkey, const secp256k1_xonly_pubkey *internal_pubkey, const unsigned char *tweak32) {
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(internal_pubkey != NULL);
+    ARG_CHECK(output_pubkey != NULL);
+    ARG_CHECK(tweak32 != NULL);
+
+    *output_pubkey = *(secp256k1_pubkey *)internal_pubkey;
+    return secp256k1_ec_pubkey_tweak_add(ctx, output_pubkey, tweak32);
+}
+
+int secp256k1_xonly_pubkey_tweak_verify(const secp256k1_context* ctx, const secp256k1_pubkey *output_pubkey, const secp256k1_xonly_pubkey *internal_pubkey, const unsigned char *tweak32) {
```

Comment:
is this function really needed? (i.e. the caller can do exactly the same thing by himself)

Reply:
And I don't remember if there's a promise that two `secp256k1_pubkey` instances can be compared.
remember I had a conversation about that with @sipa on IRC a while ago.

Reply:
Before the caller can do exactly the same thing they apparently would need to figure out how to compare pubkeys. On the other hand the taproot branch doesn't use it because it doesn't have to parse the serialized output pubkey in the first place and therefore compare the serialized versions.

Comparing via memcmp should be fine as the field elements are normalized when converting from group elements and parsing and it's used in tests all over the codebase. But this is not documented or immediately obvious and therefore not ideal. Perhaps it's time to finally add a group element comparison function.

Reply:
@sipa uses it in the taproot implementaion as he mentions below :)

Reply:
FYI it's documented that you *should* serialize `secp256k1_pubkey` before comparing https://github.com/bitcoin-core/secp256k1/blob/master/include/secp256k1.h#L64